### ISSUE TITLE
Reduce memory access and process CPU Col2im channels in parallel

### ISSUE THREAD

**Issue by Hosi121** [pytorch/pytorch#197789]
> English is not my first language, so I used AI to help write this post.
>
> CPU Col2im processes NCHW channels in a serial loop. Each channel writes to a separate output region. The patch uses `at::parallel_for` across channels for larger outputs. When tiles fully cover the output without overlap, it writes each value once. It also moves width checks out of the inner loop. The addition order and signed-zero behavior are preserved.
>
> Local FP32 Fold measurements with four threads:
>
> | FP32 Fold case, four threads | Baseline | Candidate |
> |---|---:|---:|
> | 16 output channels, 368x368, kernel/stride 2x2 | 1.546 ms | 0.108 ms |
> | 16 output channels, 256x384, kernel/stride 2x3 | 1.151 ms | 0.121 ms |
> | 16 output channels, 64x64, kernel 3x3, stride 1, padding 1 | 0.357 ms | 0.032 ms |
>
> Both versions ran as C++ extensions with the same registration and allocation path. The baseline source is `f9db06449f06a0cf6a5ff94357ae2200a9be95ac`. The host is an Intel Core Ultra 7 255H with WSL2 and GCC 13.3. The table gives medians from three fresh processes, with the run order changed between rounds.
>
> This runs the rectangular case on the installed PyTorch:
>
> ```python
> import torch
> from torch.utils.benchmark import Timer
>
> x = torch.randn(1, 16 * 2 * 3, 128 * 128)
> for threads in (1, 4):
>     timer = Timer(
>         "torch.nn.functional.fold(x, (256, 384), (2, 3), stride=(2, 3))",
>         globals={"torch": torch, "x": x},
>         num_threads=threads,
>     )
>     print(threads, timer.blocked_autorange(min_run_time=1))
> ```
>
> The outputs matched bit for bit in 864 cases across seven data types and one or four threads. Gradient checks passed with the CPU kernel replaced within the test process. The extracted functions also passed ASan/UBSan. A full PyTorch build and upstream regression tests are still needed.
>
> The change is limited to `aten/src/ATen/native/im2col.h`. Is this scope suitable for a PR?

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by frost-intel** [comment]
@Hosi121 We would welcome a perf-related PR like this, feel free to open a PR.
